File: apps/ios/src/playback/sources/playbackRequestPolicy.ts
import type {
    MediaSource,
    PlaybackPlatform,
    PlaybackRequestPolicy
} from './playbackTypes.ts';

const DEFAULT_POLICY: PlaybackRequestPolicy = {
    enableDirectPlay: true,
    enableDirectStream: true,
    allowVideoStreamCopy: true,
    allowAudioStreamCopy: true
};

const IOS_DISCOVERY_POLICY: PlaybackRequestPolicy = {
    ...DEFAULT_POLICY,
    enableDirectStream: false
};

const IOS_HLS_POLICY: PlaybackRequestPolicy = {
    enableDirectPlay: false,
    enableDirectStream: false,
    allowVideoStreamCopy: true,
    allowAudioStreamCopy: true
};

export function createPlaybackRequestPolicy(
    platform: PlaybackPlatform,
    mediaSource: MediaSource | null = null
): PlaybackRequestPolicy {
    if (platform === 'android') return DEFAULT_POLICY;
    if (platform !== 'ios') throw new Error(`unsupported playback platform: ${platform}`);
    return mediaSource?.SupportsDirectPlay === false
        ? IOS_HLS_POLICY
        : IOS_DISCOVERY_POLICY;
}

File: apps/android-tv/feature/player/src/main/kotlin/app/homeflix/tv/feature/player/PlaybackRequestPolicy.kt
package app.homeflix.tv.feature.player

data class PlaybackRequestPolicy(
    val enableDirectPlay: Boolean,
    val enableDirectStream: Boolean,
    val enableTranscoding: Boolean,
    val allowVideoStreamCopy: Boolean,
    val allowAudioStreamCopy: Boolean,
)

fun androidTvPlaybackPolicy(mediaSource: MediaSourceDto? = null): PlaybackRequestPolicy =
    PlaybackRequestPolicy(
        enableDirectPlay = mediaSource == null || mediaSource.isRemote == false && mediaSource.supportsDirectPlay,
        enableDirectStream = false,
        enableTranscoding = true,
        allowVideoStreamCopy = true,
        allowAudioStreamCopy = true,
    )
